api/revoke.go

```diff
@@ -0,0 +1,137 @@
+package api
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/smallstep/certificates/authority"
+	"golang.org/x/crypto/ocsp"
+)
+
+// RevocationReasonCodes is a map between string reason codes
+// to integers as defined in RFC 5280
+var RevocationReasonCodes = map[string]int{
+	"unspecified":          ocsp.Unspecified,
+	"keycompromise":        ocsp.KeyCompromise,
+	"cacompromise":         ocsp.CACompromise,
+	"affiliationchanged":   ocsp.AffiliationChanged,
+	"superseded":           ocsp.Superseded,
+	"cessationofoperation": ocsp.CessationOfOperation,
+	"certificatehold":      ocsp.CertificateHold,
+	"removefromcrl":        ocsp.RemoveFromCRL,
+	"privilegewithdrawn":   ocsp.PrivilegeWithdrawn,
+	"aacompromise":         ocsp.AACompromise,
+}
+
+// ReasonStringToCode tries to convert a reason string to an integer code
+func ReasonStringToCode(reason string) (int, error) {
+	// default to 0
+	if reason == "" {
+		return 0, nil
+	}
+
+	code, found := RevocationReasonCodes[strings.ToLower(reason)]
+	if !found {
+		return 0, errors.Errorf("unrecognized revocation reason '%s'", reason)
+	}
+
+	return code, nil
+}
+
+// RevokeResponse is the response object that returns the health of the server.
+type RevokeResponse struct {
+	Status string `json:"status"`
+}
+
+// RevokeRequest is the request body for a revocation request.
+type RevokeRequest struct {
+	Serial     string `json:"serial"`
+	OTT        string `json:"ott"`
+	Reason     string `json:"reason"`
+	Passive    bool   `json:"passive"`
+	CRL        bool   `json:"passive"`
+	OCSP       bool   `json:"passive"`
+	ReasonCode int
+	RTS        authority.RevocationTypeSelection
+}
+
+// Validate checks the fields of the RevokeRequest and returns nil if they are ok
+// or an error if something is wrong.
+func (r *RevokeRequest) Validate() (err error) {
+	if r.Serial == "" {
+		return BadRequest(errors.New("missing serial"))
+	}
+	if r.ReasonCode, err = ReasonStringToCode(r.Reason); err != nil {
+		return BadRequest(err)
+	}
```

Comment:
As I said before, we can receive a reason code and translate that to a string.

Reply:
Alright, so let's just receive the code and not do any conversion to string. I'll just verify that the code is within bounds. From the cli I'll allow either a `--reasonCode` or a `--reason` and do the conversion on the CLI side, sound good?

Reply:
I think it makes sense also to support an optional human reason too and store in the database.
But the only data that should be checked it would be the reasonCode.

Reply:
The api supports two reason args; `reason` (a user supplied string that can be anything), a `reasonCode` which must be one of the OCSP.Reason integers 0-10.